### PR TITLE
Pysat int2bool

### PR DESCRIPTION
**Pull request by sourdough-bread**
Improved PySAT `make_cnf()`: 
- Decoupling of cnf encoding of the constraints allowing it to be reused by `__add__()` function.
- COrrected bug where variables of newly added constraints are not added to user_vars
- Some initial documentation of the process int->bool model translation

TODO:
- Handle linear constraints by mapping with [PySAT's Pseudo-Boolean library](https://pysathq.github.io/docs/html/api/pb.html)

### CODE REVIEW

File: cpmpy/transformations/to_bool.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    elif isinstance(constraint, (list, NDVarArray)):
+        for con in constraint:
+            bool_constraints += to_bool_constraint(con, ivarmap)
+
+    # CASE 4: base comparison constraints + ensure only handling what it can
+    elif isinstance(constraint, Comparison) and all(is_int(arg) or isinstance(arg, (_IntVarImpl, _BoolVarImpl)) for arg in constraint.args) :
+        bool_constraints += to_unit_comparison(constraint, ivarmap)
+
+    # CASE 5: global constraints
+    elif isinstance(constraint, (AllDifferent, AllEqual, Circuit, Table)):
```
Comment (tias) [inline cpmpy/transformations/to_bool.py:167]: The way I see it is that, we want to replace in every constraint the intvar with its linear expression over Booleans.

Only, we also want to catch the special cases where it is a direct Comparison operator, and in that case replace the comparison by a direct expression on the Bools?

So, in any case, we should only 'do' something with constraints that contain an intvar, all others just pass through...
and the default should be to replace the intvar by a linexpr/wsum.
So then the question is, is this before or after flattening... I would say before!

**Review by tias — COMMENTED** [review]
Looks good, see comments for open points

File: cpmpy/model.py [hunk 2]
```diff
@@ -28,11 +28,18 @@
         Model
 """
 import numpy as np
+from cpmpy.expressions.variables import _BoolVarImpl
+from cpmpy.transformations.get_variables import get_variables_model
+from cpmpy.transformations.flatten_model import flatten_model
+
+from cpmpy.transformations.to_bool import intvar_to_boolvar, to_bool_constraint
```
[the remark this answers is not in the document]
Comment (sourdough-bread) [reply cpmpy/model.py:35]: Removed

File: cpmpy/model.py [hunk 3]
```diff
@@ -27,7 +27,6 @@
 
         Model
 """
-import numpy as np
```
[the remark this answers is not in the document]
Comment (tias) [reply cpmpy/model.py:30]: so all changes to this file should go

File: cpmpy/transformations/get_variables.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 import warnings # for deprecation warning
 from ..expressions.core import Expression
 from ..expressions.variables import _NumVarImpl,NegBoolView
-from ..expressions.utils import is_any_list
+from ..expressions.utils import is_any_list, is_int
```
Comment (tias) [inline cpmpy/transformations/get_variables.py:9]: no cosmetic

File: tests/test_int2bool_pysat.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+            iv1 > 6
+        )
+
+        s = CPM_pysat(m)
+        s.solve()
+
+        s += iv2 < 4
+        s.solve()
+    
+    # abs() not implemented yet
```
Comment (tias) [inline tests/test_int2bool_pysat.py:34]: its very good to add test cases too; lots of things here in comments though, remove or activate?

File: tests/test_model.py [hunk 6]
```diff
@@ -3,10 +3,7 @@
 import os
 from os.path import join
 
-from numpy import logaddexp
```
Comment (tias) [inline tests/test_model.py:6]: good catch, please commit these changes on master directly

**Comment by sourdough-bread** [comment]
Ready for reviewing and merging.

**Pull request merged**